src/main/java/net/openhft/chronicle/testframework/internal/VanillaExceptionTracker.java

```diff
@@ -0,0 +1,128 @@
+                                   @NotNull final Predicate<T> ignorePredicate) {
+        this(messageExtractor, throwableExtractor, resetRunnable, exceptions, ignorePredicate, String::valueOf);
+    }
+
+    public VanillaExceptionTracker(@NotNull final Function<T, String> messageExtractor,
+                                   @NotNull final Function<T, Throwable> throwableExtractor,
+                                   @NotNull final Runnable resetRunnable,
+                                   @NotNull final Map<T, Integer> exceptions,
+                                   @NotNull final Predicate<T> ignorePredicate,
+                                   @NotNull final Function<T, String> exceptionRenderer) {
+        this.messageExtractor = messageExtractor;
+        this.throwableExtractor = throwableExtractor;
+        this.resetRunnable = resetRunnable;
+        this.exceptions = exceptions;
+        this.ignorePredicate = ignorePredicate;
+        this.exceptionRenderer = exceptionRenderer;
+    }
+
+    @Override
+    public void expectException(String message) {
+        expectException(k -> contains(messageExtractor.apply(k), message) || (throwableExtractor.apply(k) != null && contains(throwableExtractor.apply(k).getMessage(), message)), message);
+    }
+
+    @Override
+    public void expectException(Predicate<T> predicate, String description) {
+        checkFinalised();
+        expectedExceptions.put(predicate, description);
+    }
+
+    @Override
+    public void ignoreException(String message) {
+        ignoreException(k -> contains(messageExtractor.apply(k), message) || (throwableExtractor.apply(k) != null && contains(throwableExtractor.apply(k).getMessage(), message)), message);
+    }
+
+    @Override
+    public void ignoreException(Predicate<T> predicate, String description) {
+        checkFinalised();
+        ignoredExceptions.put(predicate, description);
+    }
+
+    @Override
+    public void checkExceptions() {
+        checkFinalised();
+        finalised = true;
+        for (Map.Entry<Predicate<T>, String> expectedException : expectedExceptions.entrySet()) {
+            if (!exceptions.keySet().removeIf(expectedException.getKey()))
+                throw new AssertionError("No error for " + expectedException.getValue());
+        }
+        expectedExceptions.clear();
+        for (Map.Entry<Predicate<T>, String> ignoredException : ignoredExceptions.entrySet()) {
+            if (exceptions.keySet().removeIf(ignoredException.getKey()))
+                LOGGER.debug("Ignored {}", ignoredException.getValue());
+        }
+        ignoredExceptions.clear();
+
+        if (hasExceptions()) {
+            dumpException();
+
+            final String msg = exceptions.size() + " exceptions were detected: " + exceptions.keySet().stream().map(messageExtractor::apply).collect(Collectors.joining(", "));
+            throw new AssertionError(msg);
+        }
+        resetRunnable.run();
```

Comment:
this won't run if an AssertionError is thrown, but that's game over then anyway so fine to leave here I think.